File: java-selenium-tests/src/test/java/com/accessflow/SdkAuditBreadcrumbNavigationTest.java
package com.accessflow;

import com.acsbe.accessflow.AccessFlowSDK;
import com.acsbe.accessflow.AccessFlowTeardown;
import com.acsbe.accessflow.SeleniumDriver;
import org.junit.jupiter.api.*;
import org.openqa.selenium.*;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Breadcrumb Navigation with Accessibility Audits (Selenium).
 * Mirrors Node sdk-audit-breadcrumb-navigation.test.js.
 */
public class SdkAuditBreadcrumbNavigationTest {

    private WebDriver driver;

    @BeforeEach
    void setup() {
        driver = DriverHelper.createDriver();
    }

    @AfterEach
    void teardown() {
        if (driver != null) driver.quit();
    }

    @AfterAll
    static void finalizeReports() {
        LocalReportCollector.write();
        AccessFlowTeardown.finalizeReports();
    }

    private AccessFlowSDK createSdk() {
        String apiKey = DriverHelper.getApiKey();
        return apiKey != null && !apiKey.isEmpty()
                ? new AccessFlowSDK(new SeleniumDriver(driver), apiKey)
                : new AccessFlowSDK(new SeleniumDriver(driver));
    }

    @Test
    @DisplayName("Navigate from failure to success page via breadcrumb and audit both")
    void testNavigateFailureToSuccess() {
        DriverHelper.navigateTo(driver, "/graphics/background-images_failure");
        DriverHelper.sleep(1000);
        assertTrue(driver.getCurrentUrl().contains("background-images_failure"));

        AccessFlowSDK sdk = createSdk();
        Map<String, Object> failureReport = sdk.audit();
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), failureReport);
        LocalReportCollector.add(driver.getCurrentUrl(), failureReport);

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        WebElement breadcrumbSelect = wait.until(
                ExpectedConditions.elementToBeClickable(By.cssSelector("[aria-label='Select variant type']")));
        assertTrue(breadcrumbSelect.isDisplayed());
        DriverHelper.scrollIntoView(driver, breadcrumbSelect);
        breadcrumbSelect.click();
        DriverHelper.sleep(500);

        WebElement successOption = wait.until(
                ExpectedConditions.elementToBeClickable(By.xpath("//li[@role='option' and text()='Success']")));
        DriverHelper.scrollIntoView(driver, successOption);
        successOption.click();
        DriverHelper.sleep(1000);

        assertTrue(driver.getCurrentUrl().contains("background-images_success"));

        Map<String, Object> successReport = sdk.audit();
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), successReport);
        LocalReportCollector.add(driver.getCurrentUrl(), successReport);

        assertNotNull(failureReport);
        assertNotNull(successReport);
    }

    @Test
    @DisplayName("Navigate from success back to failure using breadcrumb")
    void testNavigateSuccessToFailure() {
        DriverHelper.navigateTo(driver, "/graphics/background-images_success");
        DriverHelper.sleep(1000);
        assertTrue(driver.getCurrentUrl().contains("background-images_success"));

        AccessFlowSDK sdk = createSdk();
        Map<String, Object> successAudit = sdk.audit();
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), successAudit);
        LocalReportCollector.add(driver.getCurrentUrl(), successAudit);

        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(10));
        WebElement breadcrumbSelect = wait.until(
                ExpectedConditions.elementToBeClickable(By.cssSelector("[aria-label='Select variant type']")));
        assertTrue(breadcrumbSelect.isDisplayed());
        DriverHelper.scrollIntoView(driver, breadcrumbSelect);
        breadcrumbSelect.click();
        DriverHelper.sleep(500);

        WebElement failureOption = wait.until(
                ExpectedConditions.elementToBeClickable(By.xpath("//li[@role='option' and text()='Failure']")));
        DriverHelper.scrollIntoView(driver, failureOption);
        failureOption.click();
        DriverHelper.sleep(1000);

        assertTrue(driver.getCurrentUrl().contains("background-images_failure"));

        Map<String, Object> failureAudit = sdk.audit();
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), failureAudit);
        LocalReportCollector.add(driver.getCurrentUrl(), failureAudit);

        assertNotNull(successAudit);
        assertNotNull(failureAudit);
    }

    @Test
    @DisplayName("Audit multiple rule pages via breadcrumb navigation")
    void testAuditMultipleRulePages() {
        List<Map<String, Object>> auditResults = new ArrayList<>();

        String[][] rules = {
                {"Background Images", "/graphics/background-images"},
                {"Alt Text", "/graphics/alt-text"},
                {"Decorative Content", "/graphics/decorative-content"}
        };

        for (String[] rule : rules) {
            DriverHelper.navigateTo(driver, rule[1] + "_failure");
            DriverHelper.sleep(800);

            AccessFlowSDK sdk = createSdk();
            Map<String, Object> failureAudit = sdk.audit();
            AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), failureAudit);
            LocalReportCollector.add(driver.getCurrentUrl(), failureAudit);

            boolean successAuditDone = false;
            try {
                WebElement breadcrumbSelect = driver.findElement(
                        By.cssSelector("[aria-label='Select variant type']"));
                if (breadcrumbSelect.isDisplayed()) {
                    DriverHelper.scrollIntoView(driver, breadcrumbSelect);
                    breadcrumbSelect.click();
                    DriverHelper.sleep(300);

                    WebElement successOption = driver.findElement(
                            By.xpath("//li[@role='option' and text()='Success']"));
                    if (successOption.isDisplayed()) {
                        DriverHelper.scrollIntoView(driver, successOption);
                        successOption.click();
                        DriverHelper.sleep(800);

                        Map<String, Object> successAudit = sdk.audit();
                        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), successAudit);
                        LocalReportCollector.add(driver.getCurrentUrl(), successAudit);
                        successAuditDone = successAudit != null;
                    }
                }
            } catch (NoSuchElementException ignored) {
                System.out.println("Breadcrumb not available for " + rule[0] + ", skipping");
            }

            auditResults.add(Map.of(
                    "rule", rule[0],
                    "failureAudit", failureAudit != null,
                    "successAudit", successAuditDone
            ));
        }

        assertTrue(auditResults.size() > 0);
        assertTrue(auditResults.stream().allMatch(r ->
                (boolean) r.get("failureAudit") && (boolean) r.get("successAudit")));
    }
}

File: selenium-test-suite/tests/sdk-audit-breadcrumb-navigation.test.js
/**
 * Breadcrumbs with Accessibility Audits (Selenium)
 *
 * Mirrors the Playwright breadcrumb-toggle-audit.spec.js using Selenium WebDriver.
 */

const { By, until } = require('selenium-webdriver');
const { AccessFlowSDK, SeleniumDriver } = require('@acsbe/accessflow-sdk');
const { createDriver, navigateTo, scrollIntoView } = require('../helpers/driver');

AccessFlowSDK.init({
  apiKey:
    process.env.AF_NODE_PACKAGE_KEY ||
    process.env.AF_Node_Package_Key ||
    'flow-1Rne8TWPwxRONacpdjA000gvFjBP5bY8fv',
});

let driver;

beforeAll(async () => {
  driver = await createDriver();
});

afterAll(async () => {
  if (driver) await driver.quit();
});

describe('Breadcrumbs with Accessibility Audits', () => {
  test('should navigate from failure to success page using breadcrumb dropdown and audit both', async () => {
    await navigateTo(driver, '/graphics/background-images_failure');
    await driver.sleep(1000);

    let currentUrl = await driver.getCurrentUrl();
    expect(currentUrl).toContain('background-images_failure');

    const sdk = new AccessFlowSDK(new SeleniumDriver(driver));

    console.log('Running audit on FAILURE page...');
    const failureReport = await sdk.audit();
    console.log(
      'Failure page audit completed:',
      failureReport ? 'Success' : 'No report',
    );

    // Open the breadcrumb dropdown (MUI Select) — scroll into view so it's clickable
    const breadcrumbSelect = await driver.wait(
      until.elementLocated(By.css('[aria-label="Select variant type"]')),
      10000,
    );

    expect(await breadcrumbSelect.isDisplayed()).toBe(true);
    await scrollIntoView(driver, breadcrumbSelect);

    console.log('Opening breadcrumb dropdown...');
    await breadcrumbSelect.click();
    await driver.sleep(500);

    // Select the "Success" option — scroll into view if needed
    console.log('Selecting SUCCESS option...');
    const successOption = await driver.wait(
      until.elementLocated(By.xpath('//li[@role="option" and text()="Success"]')),
      5000,
    );
    await scrollIntoView(driver, successOption);
    await successOption.click();

    await driver.sleep(1000);

    currentUrl = await driver.getCurrentUrl();
    expect(currentUrl).toContain('background-images_success');
    console.log('Successfully navigated to SUCCESS page');

    console.log('Running audit on SUCCESS page...');
    const successReport = await sdk.audit();
    console.log(
      'Success page audit completed:',
      successReport ? 'Success' : 'No report',
    );

    expect(failureReport).toBeDefined();
    expect(successReport).toBeDefined();
  });

  test('should navigate from success back to failure using breadcrumb', async () => {
    await navigateTo(driver, '/graphics/background-images_success');
    await driver.sleep(1000);

    let currentUrl = await driver.getCurrentUrl();
    expect(currentUrl).toContain('background-images_success');

    const sdk = new AccessFlowSDK(new SeleniumDriver(driver));

    console.log('Auditing SUCCESS page...');
    const successAudit = await sdk.audit();
    console.log('Success audit:', successAudit ? 'Complete' : 'No report');

    const breadcrumbSelect = await driver.wait(
      until.elementLocated(By.css('[aria-label="Select variant type"]')),
      10000,
    );

    expect(await breadcrumbSelect.isDisplayed()).toBe(true);
    await scrollIntoView(driver, breadcrumbSelect);

    console.log('Opening breadcrumb dropdown...');
    await breadcrumbSelect.click();
    await driver.sleep(500);

    console.log('Selecting FAILURE option...');
    const failureOption = await driver.wait(
      until.elementLocated(By.xpath('//li[@role="option" and text()="Failure"]')),
      5000,
    );
    await scrollIntoView(driver, failureOption);
    await failureOption.click();

    await driver.sleep(1000);

    currentUrl = await driver.getCurrentUrl();
    expect(currentUrl).toContain('background-images_failure');
    console.log('Successfully navigated to FAILURE page');

    console.log('Auditing FAILURE page...');
    const failureAudit = await sdk.audit();
    console.log('Failure audit:', failureAudit ? 'Complete' : 'No report');

    expect(successAudit).toBeDefined();
    expect(failureAudit).toBeDefined();
  });

  test('should audit multiple rule pages via breadcrumb navigation', async () => {
    const auditResults = [];

    const rules = [
      { name: 'Background Images', path: '/graphics/background-images' },
      { name: 'Alt Text', path: '/graphics/alt-text' },
      { name: 'Decorative Content', path: '/graphics/decorative-content' },
    ];

    for (const rule of rules) {
      await navigateTo(driver, `${rule.path}_failure`);
      await driver.sleep(800);

      const sdk = new AccessFlowSDK(new SeleniumDriver(driver));

      console.log(`\nAuditing ${rule.name} - FAILURE...`);
      const failureAudit = await sdk.audit();

      try {
        const breadcrumbSelect = await driver.findElement(
          By.css('[aria-label="Select variant type"]'),
        );

        if (await breadcrumbSelect.isDisplayed()) {
          await scrollIntoView(driver, breadcrumbSelect);
          await breadcrumbSelect.click();
          await driver.sleep(300);

          const successOption = await driver.findElement(
            By.xpath('//li[@role="option" and text()="Success"]'),
          );

          if (await successOption.isDisplayed()) {
            await scrollIntoView(driver, successOption);
            await successOption.click();
            await driver.sleep(800);

            console.log(`Auditing ${rule.name} - SUCCESS...`);
            const successAudit = await sdk.audit();

            auditResults.push({
              rule: rule.name,
              failureAudit: !!failureAudit,
              successAudit: !!successAudit,
            });

            console.log(`${rule.name}: Both audits completed`);
          }
        }
      } catch {
        console.log(`Breadcrumb not available for ${rule.name}, skipping`);
      }
    }

    console.log('\nAudit Results Summary:');
    auditResults.forEach((result) => {
      console.log(
        `   ${result.rule}: Failure ${result.failureAudit ? 'pass' : 'fail'}, Success ${result.successAudit ? 'pass' : 'fail'}`,
      );
    });

    expect(auditResults.length).toBeGreaterThan(0);
    expect(auditResults.every((r) => r.failureAudit && r.successAudit)).toBe(
      true,
    );
  });
});
